### PR TITLE
Fixed cached schemas management in RefResolver.

### PR DESCRIPTION
**Pull request by Alex-PK**
Hi,

while trying out your package today I discovered that schema caching was not working very well, as the cache was not passed by reference.

The base draft-04 schema from json-schema.org was being downloaded multiple times (as it has a cyclic reference to itself), and the download hung on the second try, maybe due to a flaky connection on my side.

This patch "centralizes" the schemas cache in the RefResolver class into a private array, avoiding the need to pass it around in the functions calls. As they are all private, this shouldn't impact any other part of the system.

All the tests are passing.

Cheers.

### CODE REVIEW

File: src/JsonSchema/RefResolver.php [hunk 1]
```diff
@@ -30,6 +30,9 @@ class RefResolver
     /** @var UriResolverInterface */
     private $uriResolver;
 
+    /** @var JsonPointer[]  */
+    private $cache = [];
```
Comment (bighappyface) [inline src/JsonSchema/RefResolver.php]: Please use `array()` here for PHP 5.3 support

Comment (Alex-PK) [inline src/JsonSchema/RefResolver.php]: Sorry, I forgot about it. :)

**Comment by jojo1981** [comment]
Now the refResolver is not stateless anymore. The file caching will now be done in the RefResolver and will be persisted. So maybe add a public method to clear the cache?

**Comment by jojo1981** [comment]
By the way the `UriRetriever` is also caching the files? So why this needs to be done twice?

**Comment by Alex-PK** [comment]
Sorry, I forgot a [] in the cache cleaning function (I am too much used to it... :) )

**Comment by bighappyface** [comment]
@Alex-PK thanks for your patience on this. Would you mind rebasing your work and squashing it down to a single commit?

@jojo1981 are you good with this?

**Comment by jojo1981** [comment]
@bighappyface I think the **RefResolver** should be stateless or a **clearCache** method should be implemented to clear the cache.

**Comment by Alex-PK** [comment]
The clearCache method has been implemented in f5269bd. I am now studying how to rebase and squash without breaking anything, as I have never done it before. :)

**Comment by Alex-PK** [comment]
Squashed and rebased. Sorry it took me so long.

**Comment by jojo1981** [comment]
@Alex-PK Because the **RefResolver** is not stateless anymore I think the documentation needs to be updated and explained the clearCache method should be used to clear the cache.

**Comment by Alex-PK** [comment]
There is no documentation... :)
What should I update? The README.md?

**Comment by jojo1981** [comment]
@Alex-PK Yes.

**Comment by bighappyface** [comment]
@Alex-PK I am glad you got to learn about rebasing and squashing on this contribution.

I am good with the rebase but want to point out that your git user configuration is probably off.

The single commit in this PR is authored by you yet submitted as an identity different than your github account:

https://github.com/justinrainbow/json-schema/pull/263/commits (shows as two people in the message)

You can see the author info using the "patch" URL: https://patch-diff.githubusercontent.com/raw/justinrainbow/json-schema/pull/263.patch

**Comment by Alex-PK** [comment]
@bighappyface Sorry, that's probably because I worked on the patch both on my work PC and my home one, and they have different configuration (email address) but with the same GitHub account.

I now updated the README.md but am unable to squash it again. When I use `git rebase -i` I get a list of all the changes in master, and if I squash the last commit I get a "could not apply" error.

**Comment by bighappyface** [comment]
@Alex-PK try the following:

```sh
# Checkout master
git checkout master
# Bring local master up to date
#  (assuming https://github.com/justinrainbow/json-schema is set to git remote "upstream")
git pull upstream master
# Checkout working branch
git checkout fix-resolver-cache
# Rebase working branch on master branch
git rebase upstream/master
# Go into interactive rebase to squash/fixup commits and isolate the last two commits
git rebase -i HEAD~2
# Use your system text editor to complete the rebase/squash and push to your fork
```

I am by no means a git expert but those steps should help you get to the right spot. Also, you should consider [amending commits](https://www.atlassian.com/git/tutorials/rewriting-history/git-commit--amend) to keep one rolling commit in a branch while you work on it.

**Comment by Alex-PK** [comment]
I think I made a mess worse than the un-rebased code. :/

**Comment by bighappyface** [comment]
Hey @Alex-PK I this PR may not be necessary anymore:

https://github.com/justinrainbow/json-schema/pull/290

Forgive me for not diving in deeper but are both PRs going for the same thing? If so, please close this PR and my apologies for taking a later PR over this one.

**Pull request closed without merging**